### PR TITLE
Add F# Compiler and mono-complete to dotnet

### PR DESCRIPTION
**Pull request by JesterOrNot**

### CODE REVIEW

File: dotnet/Dockerfile [hunk 1]
```diff
@@ -7,3 +7,7 @@ RUN wget -q https://packages.microsoft.com/config/ubuntu/19.04/packages-microsof
     add-apt-repository universe && \
     apt-get update && apt-get -y -o APT::Install-Suggests="true" install dotnet-sdk-3.1 && \
     rm -rf /var/lib/apt/lists/*;
+RUN sudo apt-get update \
+    && sudo apt-get install -y \
+        fsharp \
+        mono-complete
```
Comment (jankeromnes) [inline dotnet/Dockerfile]: Please include these packages in the RUN instruction above:

```suggestion
    apt-get update && \
    apt-get -y -o APT::Install-Suggests="true" install dotnet-sdk-3.1 && \
    apt-get install -y fsharp mono-complete && \
    rm -rf /var/lib/apt/lists/*
```

File: dotnet/Dockerfile [hunk 2]
```diff
@@ -5,5 +5,7 @@ USER root
 RUN wget -q https://packages.microsoft.com/config/ubuntu/19.04/packages-microsoft-prod.deb && \
     dpkg -i packages-microsoft-prod.deb && rm -rf packages-microsoft-prod.deb && \
     add-apt-repository universe && \
-    apt-get update && apt-get -y -o APT::Install-Suggests="true" install dotnet-sdk-3.1 && \
-    rm -rf /var/lib/apt/lists/*;
+    apt-get update && \
+    apt-get -y -o APT::Install-Suggests="true" install dotnet-sdk-3.1 && \
+    apt-get install -y fsharp mono-complete && \
```
Comment (jankeromnes) [inline dotnet/Dockerfile]: Let's try combining the two `apt-get install`s:

```suggestion
    apt-get -y -o APT::Install-Suggests="true" install \
        dotnet-sdk-3.1 \
        fsharp \
        mono-complete && \
```

File: dotnet-vnc/Dockerfile [hunk 3]
```diff
@@ -6,4 +6,8 @@ RUN wget -q https://packages.microsoft.com/config/ubuntu/18.04/packages-microsof
     dpkg -i packages-microsoft-prod.deb && rm -rf packages-microsoft-prod.deb && \
     add-apt-repository universe && \
     apt-get update && apt-get -y -o APT::Install-Suggests="true" install dotnet-sdk-2.2 && \
```
Comment (jankeromnes) [inline dotnet-vnc/Dockerfile]: Oops, it seems like I forgot to update `dotnet-vnc/Dockerfile` too when I updated `dotnet/Dockerfile` to .NET Core 3.1 in 3f81f46b3a46721cd39739dcff427ec76d637bd4!

See also https://github.com/gitpod-io/workspace-images/pull/174

**Comment by jankeromnes** [comment]
`gitpod/workspace-gecko` build failed with:

```
Your version of Rust (1.41.0) is too old.
Found rustup. Will try to upgrade.

We attempted to upgrade Rust to a modern version (1.41.1 or newer).
However, you appear to still have version 1.41.0.

It's possible rustup failed. It's also possible the new Rust is not being
installed in the search path for this shell. Try creating a new shell and
run this bootstrapper again.

If this continues to fail and you are sure you have a modern Rust on your
system, ensure it is on the $PATH and try again. If that fails, you'll need to
install Rust manually.

We recommend the installer from https://rustup.rs/ for installing Rust,
but you may be able to get a recent enough version from a software install
tool or package manager on your system, or directly from https://rust-lang.org/
```

i.e. **our current Rust version is too old** (unrelated to this PR).

---

`gitpod/workspace-dotnet` and `gitpod/workspace-dotnet-vnc` failed with:

```
Please select the mail server configuration type that best meets your needs.

 No configuration:
  Should be chosen to leave the current configuration unchanged.
 Internet site:
  Mail is sent and received directly using SMTP.
 Internet with smarthost:
  Mail is received directly using SMTP or by running a utility such
  as fetchmail. Outgoing mail is sent using a smarthost.
 Satellite system:
  All mail is sent to another machine, called a 'smarthost', for delivery.
 Local only:
  The only delivered mail is the mail for local users. There is no network.

  1. No configuration  3. Internet with smarthost  5. Local only
  2. Internet Site     4. Satellite system
General type of mail configuration: 
Too long with no output (exceeded 30m0s): context deadline exceeded
```

i.e. some package wants to interactively ask for something, which isn't possible in Docker builds. We should make the install non-interactive.

**Comment by JesterOrNot** [comment]
see #176

**Comment by jankeromnes** [comment]
@JesterOrNot Thanks for upgrading the pinned Rust version!

Regarding making the `apt-get install` non-interactive, you could try using `DEBIAN_FRONTEND=noninteractive`, like in:

https://github.com/gitpod-io/workspace-images/blob/792e70941b16500d0e32cd9ec01af57b2fe11e21/full/Dockerfile#L70

or:

https://github.com/gitpod-io/workspace-images/blob/792e70941b16500d0e32cd9ec01af57b2fe11e21/full/Dockerfile#L244

**Comment by jankeromnes** [comment]
Rebased and added `DEBIAN_FRONTEND=noninteractive`.

**Comment by JesterOrNot** [comment]
Thanks!

**Comment by jankeromnes** [comment]
Failed with:

```
Errors were encountered while processing:

 /tmp/apt-dpkg-install-UR58Xv/3357-msodbcsql17_17.4.2.1-1_amd64.deb

E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by jankeromnes** [comment]
FYI, I found out about `ACCEPT_EULA=Y` being required for `msodbcsql` in https://github.com/docker-library/php/issues/424#issuecomment-319427254.

**Comment by jankeromnes** [comment]
Also, for the record, the `postfix` install issue shown above could probably also have been fixed with:

```
debconf-set-selections <<< "postfix postfix/mailname string your.hostname.com"
debconf-set-selections <<< "postfix postfix/main_mailer_type string 'Internet Site'"
```

See: https://serverfault.com/a/144010

**Comment by jankeromnes** [comment]
Both dotnet images failed with:

```
Setting up resolvconf (1.79ubuntu13) ...

/var/lib/dpkg/info/resolvconf.config: 13: /var/lib/dpkg/info/resolvconf.config: ifquery: not found

/var/lib/dpkg/info/resolvconf.config: 13: /var/lib/dpkg/info/resolvconf.config: ifquery: not found

Created symlink /etc/systemd/system/sysinit.target.wants/resolvconf.service → /lib/systemd/system/resolvconf.service.

Created symlink /etc/systemd/system/systemd-resolved.service.wants/resolvconf-pull-resolved.path → /lib/systemd/system/resolvconf-pull-resolved.path.

ln: failed to create symbolic link '/etc/resolv.conf': Device or resource busy

dpkg: error processing package resolvconf (--configure):

 installed resolvconf package post-installation script subprocess returned error exit status 1
```

**Comment by jankeromnes** [comment]
All green! 🍏 📗 💚 🥗 ✅ 🚀 

Screenshot 2020-03-06 at 12 01 39

**Pull request merged**